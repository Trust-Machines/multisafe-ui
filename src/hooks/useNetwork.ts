import {useAtom} from 'jotai';
import {networkAtom} from '../store';
import {Network} from '../store/network';

const useNetwork = (): [Network, (n: Network) => void] => {
    const [network, setNetwork_] = useAtom(networkAtom);

    const setNetwork = (n: Network) => {
        setNetwork_(n);
        localStorage.setItem('app_network', n)
    }
    return [network, setNetwork];
}

export default useNetwork;