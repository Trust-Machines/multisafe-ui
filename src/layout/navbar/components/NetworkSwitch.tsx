import React, {useState} from 'react';

import Box from '@mui/material/Box';
import Chip from '@mui/material/Chip';
import ClickAwayListener from '@mui/material/ClickAwayListener';
import {grey} from '@mui/material/colors';

import ThemedBox from '../../../components/themed-box';
import useNetwork from '../../../hooks/useNetwork';
import {useTheme} from '@mui/material';


const NetworkChip = (props: { label: string, onClick?: () => void }) => {
    return <Chip sx={{cursor: 'pointer', textTransform: 'capitalize'}} label={props.label}
                 color={props.label === 'mainnet' ? 'primary' : 'default'} onClick={props.onClick}/>;
}

export const NetworkMenu = (props: { onChange: () => void }) => {
    const [, , setNetwork] = useNetwork();

    return (
        <ThemedBox sx={{
            position: 'absolute',
            width: 'calc(100% - 20px)',
            left: '0',
            top: '60px',
        }}>
            <Box sx={{marginBottom: '6px'}}>
                <NetworkChip label='mainnet' onClick={() => {
                    setNetwork('mainnet');
                    props.onChange();
                }}/>
            </Box>
            <Box>
                <NetworkChip label='testnet' onClick={() => {
                    setNetwork('testnet');
                    props.onChange();
                }}/>
            </Box>
        </ThemedBox>
    )
}

const NetworkSwitch = () => {
    const [menu, setMenu] = useState(false);
    const [network] = useNetwork();
    const theme = useTheme();

    return (
        <ClickAwayListener onClickAway={() => {
            setMenu(false);
        }}>
            <Box sx={{
                flexShrink: 0,
                width: '100px',
                height: '100%',
                position: 'relative',
                borderRight: `2px solid ${theme.palette.divider}`,
                '&:hover': {
                    background: theme.palette.mode === 'light' ? grey[50] : grey[800]
                }
            }} onClick={() => {
                setMenu(true);
            }}>
                <Box sx={{
                    height: '100%',
                    display: 'flex',
                    alignItems: 'center',
                    justifyContent: 'center',
                    cursor: 'pointer',
                }}>
                    <NetworkChip label={network}/>
                </Box>
                {menu && <NetworkMenu onChange={() => {
                    setTimeout(() => {
                        setMenu(false);
                    }, 100);
                }}/>}
            </Box>
        </ClickAwayListener>);
}

export default NetworkSwitch;