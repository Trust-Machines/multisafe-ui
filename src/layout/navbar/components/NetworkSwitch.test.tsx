import {render} from '@testing-library/react';

import NetworkSwitch, {NetworkMenu} from './NetworkSwitch';

import useNetwork from '../../../hooks/use-network';
jest.mock('../../../hooks/use-network');

test('1 Default render', () => {
    (useNetwork as jest.Mock).mockReturnValue(['mainnet']);
    const rendered = render(<NetworkSwitch/>);
    expect(rendered.container).toMatchSnapshot()
});

test('2 Render with testnet', () => {
    (useNetwork as jest.Mock).mockReturnValue(['testnet']);
    const rendered = render(<NetworkSwitch/>);
    expect(rendered.container).toMatchSnapshot()
});

test('3 Default render NetworkMenu', () => {
    (useNetwork as jest.Mock).mockReturnValue(['testnet', {}, ()=>{}]);
    const rendered = render(<NetworkMenu onChange={()=>{}}/>);
    expect(rendered.container).toMatchSnapshot()
});