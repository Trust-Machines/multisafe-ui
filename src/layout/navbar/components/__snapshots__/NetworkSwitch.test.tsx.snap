// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 Default render 1`] = `
<div>
  <div
    class="MuiBox-root css-1n1j97o"
  >
    <div
      class="MuiBox-root css-1sfu4nm"
    >
      <div
        class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-clickable MuiChip-clickableColorPrimary MuiChip-filledPrimary css-8w4ikn-MuiButtonBase-root-MuiChip-root"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          mainnet
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`2 Render with testnet 1`] = `
<div>
  <div
    class="MuiBox-root css-1n1j97o"
  >
    <div
      class="MuiBox-root css-1sfu4nm"
    >
      <div
        class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-av5qre-MuiButtonBase-root-MuiChip-root"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          testnet
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`3 Default render NetworkMenu 1`] = `
<div>
  <div
    class="MuiBox-root css-9kanww"
  >
    <div
      class="MuiBox-root css-jmbq56"
    >
      <div
        class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorPrimary MuiChip-clickable MuiChip-clickableColorPrimary MuiChip-filledPrimary css-8w4ikn-MuiButtonBase-root-MuiChip-root"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          mainnet
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-av5qre-MuiButtonBase-root-MuiChip-root"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          testnet
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </div>
  </div>
</div>
`;
